internal/controller/s3clienthandler/s3clienthandler.go

```diff
@@ -1,55 +1,142 @@
-	return func(c *Handler) {
-		c.kubeClient = k
+	return func(h *Handler) {
+		h.kubeClient = k
 	}
 }
 
 func WithAssumeRoleArn(a *string) func(*Handler) {
-	return func(c *Handler) {
-		c.assumeRoleArn = a
+	return func(h *Handler) {
+		h.assumeRoleArn = a
 	}
 }
 
 func WithBackendStore(s *backendstore.BackendStore) func(*Handler) {
-	return func(c *Handler) {
-		c.backendStore = s
+	return func(h *Handler) {
+		h.backendStore = s
+	}
+}
+
+func WithS3Timeout(t time.Duration) func(*Handler) {
+	return func(h *Handler) {
+		h.s3Timeout = t
 	}
 }
 
 func WithLog(l logging.Logger) func(*Handler) {
-	return func(c *Handler) {
-		c.log = l
+	return func(h *Handler) {
+		h.log = l
 	}
 }
 
-func (c *Handler) GetS3Client(backendName string) (backendstore.S3Client, error) {
-	// TODO: We should only return the existing backend s3 client if the user has not
-	// specified --assume-role-arn. Otherwise, we should use the backend's STS client
-	// perform AssumeRole and use the response credentials to buld a temporary S3 client
-	// for the operation being undertaken.
-	return c.backendStore.GetBackendS3Client(backendName), nil
+func (h *Handler) GetS3Client(ctx context.Context, b *v1alpha1.Bucket, backendName string) (backendstore.S3Client, error) {
+	if h.assumeRoleArn != nil {
+		return h.createAssumeRoleS3Client(ctx, b, backendName)
+	}
+
+	return h.backendStore.GetBackendS3Client(backendName), nil
+}
+
+func (h *Handler) createAssumeRoleS3Client(ctx context.Context, b *v1alpha1.Bucket, backendName string) (backendstore.S3Client, error) {
+	roleSessionName, err := newRoleSessionNameGenerator().generate(roleSessionNamePrefix)
+	if err != nil {
+		return nil, errors.Wrap(err, errFailedToCreateAssumeRoleS3Client.Error())
+	}
+
+	copiedTags := make([]ststypes.Tag, 0)
+	if b.Spec.ForProvider.AssumeRoleTags != nil {
+		copiedTags = copySTSTags(b.Spec.ForProvider.AssumeRoleTags)
+	}
+
+	input := &sts.AssumeRoleInput{
+		RoleArn:         h.assumeRoleArn,
+		RoleSessionName: &roleSessionName,
+		Tags:            copiedTags,
+	}
+
+	stsClient := h.backendStore.GetBackendSTSClient(backendName)
+	if stsClient == nil {
+		return nil, errors.Wrap(errNoSTSClient, errFailedToCreateAssumeRoleS3Client.Error())
+	}
+
+	resp, err := rgw.AssumeRole(ctx, stsClient, input)
+	if err != nil {
+		return nil, errors.Wrap(err, errFailedToCreateAssumeRoleS3Client.Error())
+	}
+
+	if resp.Credentials == nil || resp.Credentials.AccessKeyId == nil || resp.Credentials.SecretAccessKey == nil {
+		return nil, errors.Wrap(errNoCreds, errFailedToCreateAssumeRoleS3Client.Error())
+	}
+
+	data := map[string][]byte{
+		consts.KeyAccessKey: []byte(*resp.Credentials.AccessKeyId),
+		consts.KeySecretKey: []byte(*resp.Credentials.SecretAccessKey)}
+
+	pc := &apisv1alpha1.ProviderConfig{}
+	if err := h.kubeClient.Get(ctx, types.NamespacedName{Name: backendName}, pc); err != nil {
+		return nil, errors.Wrap(err, errFailedToCreateAssumeRoleS3Client.Error())
```

Comment:
Does it make sense to handle not found error?

Reply:
I don't think it makes any difference what the error is in this case - if there is an error we just log and skip the backend. I do believe we should update the CR status though - I will open a separate PR for this soon